drydock/plugin.py

```diff
@@ -1,11 +1,179 @@
+            if template["spec"]["template"]["spec"]["containers"][0].get("command") == []:
+                # In some cases, we need to bypass the container entrypoint.
+                # Unfortunately, AFAIK, there is no way to do so in K8s manifests. So we mark
+                # some jobs with "command: []". For these jobs, the entrypoint becomes "sh -e -c".
+                # We do not do this for every job, because some (most) entrypoints are actually useful.
+                template["spec"]["template"]["spec"]["containers"][0]["command"] = shell_command
+                container_args = [render_command]
+            else:
+                container_args = shell_command + [render_command]
+
+            template["spec"]["template"]["spec"]["containers"][0]["args"] = container_args
+            template["spec"]["backoffLimit"] = 1
+            template["spec"]["ttlSecondsAfterFinished"] = 3600
+
+
+            yield serialize.dumps(template)
+
+    return response
+
+
+def get_upgrade_list():
+    """
+    Return a list of upgrade target versions based on the MIGRATE_FROM setting.
+    """
+    context = click.get_current_context().obj
+    tutor_conf = tutor_config.load(context.root)
+
+    upgrade_list = []
+    if not 'MIGRATE_FROM' in tutor_conf:
+        return upgrade_list
+    migrate_from = tutor_conf["MIGRATE_FROM"]
+    migrate_to = config['defaults']['VERSION'].split('.', maxsplit=1)[0]
+    for name, version in VERSION_LIST:
+        if migrate_from.lower() == name.lower():
+            migrate_from = version
+        if migrate_from <= version <= migrate_to:
+            upgrade_list.append(name.lower())
+    return upgrade_list
+
+
+CORE_SYNC_WAVES_ORDER: t.Dict[str, SYNC_WAVES_ORDER_ATTRS_TYPE] = {
+    "jobs": {
+        "drydock-upgrade-lms-job": 50,
+        "drydock-upgrade-cms-job": 51
+    },
+    "deployments": {
+        "lms-lifecycle-enabled": 100,
+        "cms-lifecycle-enabled": 100,
+        "lms-debug": 50,
+        "cms-debug": 50,
+    },
+    "horizontalpodautoscalers": {
+        "all": 150
+    },
+    "ingresses": {
+        "ingress-debug": 200
+    },
+}
+
+# The core sync-waves configs are added with a high priority, such that other users can override or
+# remove them.
+@SYNC_WAVES_ORDER.add(priority=tutor_hooks.priorities.HIGH)
+def _add_core_sync_waves_order(sync_waves_config: t.Dict[str, SYNC_WAVES_ORDER_ATTRS_TYPE]) -> t.Dict[str, SYNC_WAVES_ORDER_ATTRS_TYPE]:
+    sync_waves_config.update(CORE_SYNC_WAVES_ORDER)
+    return sync_waves_config
+
+
+@functools.lru_cache(maxsize=None)
+def get_sync_waves_order() -> t.Dict[str, SYNC_WAVES_ORDER_ATTRS_TYPE]:
+    """
+    This function is cached for performance.
+    """
+    return SYNC_WAVES_ORDER.apply({})
+
+
+@tutor_hooks.Actions.PLUGIN_LOADED
+def _clear_sync_waves_order_cache(_name: str) -> None:
+    """
+    Don't forget to clear cache, or we'll have some strange surprises...
+    """
+    get_sync_waves_order.cache_clear()
```

Comment:
Does this cache bring that much performance improvement?

Reply:
I don't know, it's copy-pasted from the `tutor-mfe` plugin and seems to be important

Reply:
What happens if you remove it? I tried running 'time tutor config save' with the decorator and without and didn't see much improvement.